### PR TITLE
Workflow step: Auto-commit images

### PR DESCRIPTION
**Pull request by josecelano**
New workflow step in the Gold Drawings Processing Workflow. The new step is the auto-commit for the automatically generated Base image.

### CODE REVIEW

**Comment by josecelano** [comment]
@yeraydavidrodriguez @da2ce7 

I've only implemented the test scaffolding so far. The "arrange" part. Now I have to implement the step (importing the code from [here](https://github.com/josecelano/pygithub/blob/main/src/03_sign_commit_using_the_gitpython_package.py#L221)).

This is the main file to review: https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40/files#diff-f26612711fcfde16848c91d8467da76ba292e684b7c150a14d454548e6f101b7

File: tests/test_nautilus_librarian/utils.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import subprocess  # nosec
+
+
+def debug_execute_console_command(multiline_command):
+    return execute_console_command(
+        multiline_command, print_output=True, print_command=True
+    )
+
+
+def execute_console_command(multiline_command, print_output=False, print_command=False):
```
Comment (da2ce7) [inline tests/test_nautilus_librarian/utils.py:11]: `print_output` and `print_command` are both unused.

File: tests/test_nautilus_librarian/utils.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+        if print_output:
+            print(output)
+
+        full_output += output
+
+    return full_output
+
+
+def get_fixtures_dir():
```
Comment (da2ce7) [inline tests/test_nautilus_librarian/utils.py:40]: What is the purpose of this function?

**Review by da2ce7 — CHANGES_REQUESTED** [review]
For the tests, I think that we should use the Python Unit Testing Framework:
https://docs.python.org/3/library/unittest.html

File: tests/test_nautilus_librarian/utils.py (hunk 1 above)
Comment (josecelano) [reply tests/test_nautilus_librarian/utils.py:11]: I use the function `debug_execute_console_command` whilst debugging a given test. I have to run a lot of console commands (git init, dvc add , ...) and it's pretty handy to print the full output to check what is going wrong. Maybe we can automatically change the behaviors of `debug_execute_console_command` when you run the tests with the option `pytest -s`, but I suppose you you do not always want to print all.

File: tests/test_nautilus_librarian/utils.py (hunk 2 above)
Comment (josecelano) [reply tests/test_nautilus_librarian/utils.py:40]: It's a helper function to use only in tests. It gives you the test folder where I'm going to store tests fixtures like: example images, gpg keys, etcetera. I can add this description :-) to the docstring.

**Comment by josecelano** [comment]
> For the tests, I think that we should use the Python Unit Testing Framework: https://docs.python.org/3/library/unittest.html

[Why?](https://www.pythonpool.com/python-unittest-vs-pytest/)

To be honest, I did not even look for the alternatives. Somehow I assumed that Pytest was the "modern" package.

File: tests/test_nautilus_librarian/utils.py (hunk 1 above)
Comment (da2ce7) [reply tests/test_nautilus_librarian/utils.py:11]: However both arguments, `print_output` and `print_command`, are not referenced in the function body at all...

**Comment by da2ce7** [comment]
> > For the tests, I think that we should use the Python Unit Testing Framework: https://docs.python.org/3/library/unittest.html
> 
> @da2ce7 [Why?](https://www.pythonpool.com/python-unittest-vs-pytest/)
> 
> To be honest, I did not even look for the alternatives. Somehow I assumed that Pytest was the "modern" package.

Ah... I guess you are right. :P  Pytest seems to the be the modern standard.

**Comment by josecelano** [comment]
hi @da2ce7 @yeraydavidrodriguez , it's almost finished, there are still some pending tasks:

- [ ] Refactor `auto_commit_base_images_step` function. I want to write unit tests for each workflow step. The [acceptance test](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40/files?authenticity_token=4QISolHGDCJMkeErm8soweEKYCOW5wdaUfbLaEamjdKoYqHNymJY0oi17lDL0%2Fz5Cvbz9HWuRL9xVjrqyVnkjg%3D%3D&file-filters%5B%5D=.py&hide-deleted-files=true#diff-f26612711fcfde16848c91d8467da76ba292e684b7c150a14d454548e6f101b7R20-R76) for the whole workflow is slow, although it has been a good golden test to develop this feature.
- [ ] There are still [two cases](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40/files?authenticity_token=4QISolHGDCJMkeErm8soweEKYCOW5wdaUfbLaEamjdKoYqHNymJY0oi17lDL0%2Fz5Cvbz9HWuRL9xVjrqyVnkjg%3D%3D&file-filters%5B%5D=.py&hide-deleted-files=true#diff-4bd6ed94c017bff5c96d1add061ec42315d8b01066c77e48412976de08c318f8R115) pending to define. I can't finish that part until we [define the process to update and rename a Gold image](https://github.com/Nautilus-Cyberneering/chinese-ideographs/pull/122#issuecomment-972844365) and how that affects the this workflow.

Notes:

- @yeraydavidrodriguez I've added some [dvc command wrappers](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40/files?authenticity_token=4QISolHGDCJMkeErm8soweEKYCOW5wdaUfbLaEamjdKoYqHNymJY0oi17lDL0%2Fz5Cvbz9HWuRL9xVjrqyVnkjg%3D%3D&file-filters%5B%5D=.py&hide-deleted-files=true#diff-81274a302664ff55caabe23e35ff583ceaa199e9d65170a6ecdf95498ca45a85R60). They are a temporary solution until we merge the [PR for the dvc wrapper](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/25).
- Regarding signing commits, the [repo.commit](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40/files?authenticity_token=4QISolHGDCJMkeErm8soweEKYCOW5wdaUfbLaEamjdKoYqHNymJY0oi17lDL0%2Fz5Cvbz9HWuRL9xVjrqyVnkjg%3D%3D&file-filters%5B%5D=.py&hide-deleted-files=true#diff-65e8a306606e567e028753d653a18eaa62bcf4716d5e8c0647a861f907a8f6daR12) method is using the global git user configuration. In the dev env you should already have your own signingkey (otherwise it will fail). In the `test` workflow I'm using a [marketplace action](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40/files?authenticity_token=4QISolHGDCJMkeErm8soweEKYCOW5wdaUfbLaEamjdKoYqHNymJY0oi17lDL0%2Fz5Cvbz9HWuRL9xVjrqyVnkjg%3D%3D&file-filters%5B%5D=.py&file-filters%5B%5D=.yml&hide-deleted-files=true#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R17-R31), although we could create our own GPG key importer. I've already did it [here](https://github.com/josecelano/pygithub/blob/main/src/03_sign_commit_using_the_gitpython_package.py#L125).

**Comment by josecelano** [comment]
I've done the refactor for "validate filenames action":

```
def validate_filenames(dvc_diff):
    """
    It validates all the media file names in the dvc diff.
    """
    if dvc_diff == "{}":
        return ActionResult(ResultCode.EXIT, [Message("No Gold image changes found")])

    filenames = extract_modified_media_file_list_from_dvd_diff_output(dvc_diff)

    messages = []

    for filename in filenames:
        try:
            validate_filename(filename)
            messages.append(Message(f"{filename} ✓"))
        except ValueError as error:
            return ActionResult(ResultCode.ABORT, [Message(f"{filename} ✗ {error}")])

    return ActionResult(ResultCode.CONTINUE, messages)
```

I remove the Typer dependency. Now we can have simpler units tests for each step:

```
def given_a_dvc_diff_object_it_should_validate_the_filename_of_the_new_media_files(
    temp_git_dir, temp_dvc_local_remote_storage_dir, sample_base_image_absolute_path
):
    dvc_diff_with_added_gold_image = {
        "added": [
            {"path": "data/000001/32/000001-32.600.2.tif"},
        ],
        "deleted": [],
        "modified": [],
        "renamed": [],
    }

    result = validate_filenames(compact_json(dvc_diff_with_added_gold_image))

    assert result.code == ResultCode.CONTINUE
    assert result.contains_text("000001-32.600.2.tif ✓")
```

It takes long to run the acceptance test running the console command:

```
def given_a_dvc_diff_object_it_should_validate_the_filename_of_the_new_media_files(
    temp_git_dir, temp_dvc_local_remote_storage_dir, sample_base_image_absolute_path
):
    create_initial_state(
        temp_git_dir, temp_dvc_local_remote_storage_dir, sample_base_image_absolute_path
    )

    dvc_diff_with_added_gold_image = {
        "added": [
            {"path": "data/000001/32/000001-32.600.2.tif"},
        ],
        "deleted": [],
        "modified": [],
        "renamed": [],
    }

    # Execute the workflow
    result = runner.invoke(
        app,
        ["gold-drawings-processing", compact_json(dvc_diff_with_added_gold_image)],
        env={"INPUT_GIT_REPO_DIR": str(temp_git_dir)},
    )

    assert result.exit_code == 0
    assert "000001-32.600.2.tif ✓" in result.stdout
```

It's takes some seconds to create the initial state and we have to do it for a lot of tests. We have implemented only 2 steps (actions) and it takes a couple of seconds. I'm going to change also the other action to use this approach.

**Comment by josecelano** [comment]
I think we can merge the PR as it is.

I'm going to skip the documentation for the time being until we finish the command for the workflow.

I will create new issues for the cases I've not implemented here:

- Delete Gold Image
- Modify Gold Image
- Rename Gold Image

As I mentioned [above](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40#issuecomment-996871998) those cases are still [pending to define](https://github.com/Nautilus-Cyberneering/chinese-ideographs/pull/122#issuecomment-972844365).

I will create a new issue to refactor the [gpg functions imported](https://github.com/Nautilus-Cyberneering/nautilus-librarian/pull/40/files#diff-f552afdda9db9b2fde310e69691da6b85a58cdd8876954b70c2abd8e3715a744R1-R127) into the gpg mod from the [external repo](https://github.com/josecelano/pygithub).

File: src/nautilus_librarian/mods/gpg/domain/utils.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            uid = fields[9]
+            if current_fingerprint == fingerprint:
+                name, separator, rest = uid.partition(" <")
+                email, separator, rest = rest.partition(">")
+                return name, email
+
+    return None
+
+
+def get_short_key_id_from_fingerprint(fingerprint):
```
Comment (da2ce7) [inline src/nautilus_librarian/mods/gpg/domain/utils.py:87]: Is there are particular reason we need to use `Short Key ID's` or `Long Key ID's`? Both are so short they can be easily ground.

**Review by da2ce7 — COMMENTED** [review]
Post-Merge Ack.

Good work. :) Cam.

**Pull request merged**